src/merkle/partial_mt/mod.rs

```diff
@@ -0,0 +1,270 @@
+use super::{
+    BTreeMap, BTreeSet, InnerNodeInfo, MerkleError, MerklePath, NodeIndex, Rpo256, RpoDigest, Vec,
+    Word, EMPTY_WORD,
+};
+
+#[cfg(test)]
+mod tests;
+
+// PARTIAL MERKLE TREE
+// ================================================================================================
+
+/// A partial Merkle tree with NodeIndex keys and 4-element RpoDigest leaf values.
+///
+/// The root of the tree is recomputed on each new leaf update.
+pub struct PartialMerkleTree {
+    root: RpoDigest,
+    max_depth: u8,
+    nodes: BTreeMap<NodeIndex, RpoDigest>,
+    leaves: BTreeSet<NodeIndex>,
+}
+
+impl Default for PartialMerkleTree {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl PartialMerkleTree {
+    // CONSTANTS
+    // --------------------------------------------------------------------------------------------
+
+    /// An RpoDigest consisting of 4 ZERO elements.
+    pub const EMPTY_DIGEST: RpoDigest = RpoDigest::new(EMPTY_WORD);
+
+    /// Minimum supported depth.
+    pub const MIN_DEPTH: u8 = 1;
+
+    /// Maximum supported depth.
+    pub const MAX_DEPTH: u8 = 64;
+
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns a new emply [PartialMerkleTree].
+    pub fn new() -> Self {
+        PartialMerkleTree {
+            root: Self::EMPTY_DIGEST,
+            max_depth: 0,
+            nodes: BTreeMap::new(),
+            leaves: BTreeSet::new(),
+        }
+    }
+
+    /// Returns a new [PartialMerkleTree] instantiated with leaves set as specified by the provided
+    /// entries.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - If the depth is 0 or is greater than 64.
+    /// - The number of entries exceeds the maximum tree capacity, that is 2^{depth}.
+    /// - The provided entries contain multiple values for the same key.
+    pub fn with_leaves<R, I>(entries: R) -> Result<Self, MerkleError>
+    where
+        R: IntoIterator<IntoIter = I>,
+        I: Iterator<Item = (NodeIndex, RpoDigest)> + ExactSizeIterator,
+    {
+        // create an empty tree
+        let mut tree = PartialMerkleTree::new();
+
+        // check if the number of leaves can be accommodated by the tree's depth; we use a min
+        // depth of 63 because we consider passing in a vector of size 2^64 infeasible.
+        let entries = entries.into_iter();
+        let max = (1_u64 << 63) as usize;
+        if entries.len() > max {
+            return Err(MerkleError::InvalidNumEntries(max, entries.len()));
+        }
+
+        for (node_index, rpo_digest) in entries {
+            let old_value = tree.update_leaf(node_index, rpo_digest)?;
+            if old_value != Self::EMPTY_DIGEST {
+                return Err(MerkleError::DuplicateValuesForIndex(node_index.value()));
+            }
+        }
+        Ok(tree)
+    }
```

Comment:
We should not be able to create a `PartialMerkleTree` from leaves. If we are able to do so, then it becomes a regular Merkle tree. What we want is to create a partial Merkle tree from Merkle paths (similar to [this](https://github.com/0xPolygonMiden/crypto/blob/main/src/merkle/path_set.rs#L33) method in `MerklePathSet`). I would change this to something like:
```Rust
pub fn with_paths<I>(paths: I) -> Result<Self, MerkleError>
where
    I: IntoIterator<Item = (u64, Word, MerklePath)>,
{
    ...
}
```
The main difference between this is `MerklePathSet` is that for `PartialMerkleTree` paths could be of different lengths.

In the future, we can add other constructors - but I think for now this is all we need.

Reply:
Got it, I'll create `with_paths` constructor. But maybe we should keep `with_leaves` one, because it is possible to pass "leaves" with different depths into it (so they're not actually leaves, more like outer nodes). So, for example, these two trees are equal:
```
with_paths([((2, 0), A, [(2, 1), (1, 1)]),
            ((3, 2), B, [(3, 3), (2, 0), (1, 1)]),
            ((3, 3), C, [(3, 2), (2, 0), (1, 1)]),
            ((1, 1), D, [(1, 0)]),
])
with_leaves([((2, 0), A), ((3, 2), B), ((3, 3), C), ((1, 1), D)])
```
since they both construct this tree:
```
                _______ (0, 0) _______
               /                      \
         _ (1, 0) _               _____ D _____
        /          \             /             \
    _ A _        (2, 1)       (2, 2)         (2, 3)
   /     \        /  \       /     \         /     \
(3, 0) (3, 1)    B    C   (3, 4) (3, 5)   (3, 6) (3, 7)
```
But on the other hand, Merkle paths are more common way to create trees, so, maybe, we just don't need `with_leaves` constructor.

Reply:
I think with `with_leaves()` could work too - but we need to make sure we detect missing entries. For example, for the above tree, `with_leaves([((2, 0), A), ((3, 2), B), ((3, 3), C)])` should be an error (because we don't know what is under the right subtree of the root.

We also need to make sure that no duplicate entries are provided. For example, `with_leaves([((2, 0), A), ((3, 2), B), ((3, 3), C), ((1, 1), D), ((2, 2), F)])` should be an error (because `D` is no longer a leaf. Though, I think maybe this check is already happening.

Basically, I wonder if `with_leaves()` is a bit more complicated and something we can do in a follow-up PR.